flow_runner.go

```diff
@@ -0,0 +1,174 @@
+package taskflow
+
+import (
+	"context"
+	"errors"
+	"flag"
+	"fmt"
+	"io"
+	"sort"
+	"strings"
+	"text/tabwriter"
+	"time"
+)
+
+type flowRunner struct {
+	output  io.Writer
+	params  Params
+	tasks   map[string]Task
+	verbose bool
+}
+
+// Run runs provided tasks and all their dependencies.
+// Each task is executed at most once.
+func (f *flowRunner) Run(ctx context.Context, args []string) int {
+	// parse args
+	cli := flag.NewFlagSet("", flag.ContinueOnError)
+	cli.SetOutput(f.output)
+	verbose := cli.Bool("v", false, "Verbose output: log all tasks as they are run. Also print all text from Log and Logf calls even if the task succeeds.")
+	usage := func() {
+		fmt.Fprintf(cli.Output(), "Usage: [flag(s)] task(s)\n")
+		fmt.Fprintf(cli.Output(), "Flags:\n")
+		cli.PrintDefaults()
+
+		fmt.Fprintf(cli.Output(), "Tasks:\n")
+		w := tabwriter.NewWriter(cli.Output(), 1, 1, 4, ' ', 0)
+		keys := make([]string, 0, len(f.tasks))
+		for k, task := range f.tasks {
+			if task.Description == "" {
+				continue
+			}
+			keys = append(keys, k)
+		}
+		sort.Strings(keys)
+		for _, k := range keys {
+			t := f.tasks[k]
+			fmt.Fprintf(w, "  %s\t%s\n", t.Name, t.Description)
+		}
+		if err := w.Flush(); err != nil {
+			panic(err)
+		}
+	}
+	cli.Usage = usage
+	if err := cli.Parse(args); err != nil {
+		fmt.Fprintln(cli.Output(), err)
+		return CodeInvalidArgs
+	}
+	if *verbose {
+		f.verbose = true
+	}
+
+	// parse non-flag args
```

Comment:
for me, the for loop looks a bit like magic. How about extracting it to a function?

Reply:
I think of changing `func (f *flowRunner) Run(ctx context.Context, args []string) int {` to something like `func (f *flowRunner) runFlow(ctx context.Context, args flowArgs) int {`. Maybe functional style would make this code more readable and easier to maintain. However, I will try it in separate PR, because it is not critical and it will be easier to discuss that way.